---
title: "Evaluation metrics LRGASP -- Challenge 3"
output: html_document
params:
  date: !r Sys.Date()
  output.directory: ""   #folder where the evaluation output will be saved
  Name: "Example"
  Platform: "PacBio"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
getwd()
```

```{r, echo=FALSE}
## These are a few libraries that must be loaded for the plots and also a couple of variables to make them "nice"

suppressWarnings(library(ggplot2))
suppressWarnings(library(scales))
suppressWarnings(library(knitr))


myPalette = c("#6BAED6","#FC8D59","#78C679","#EE6A50","#969696","#66C2A4", "goldenrod1", "darksalmon", "#41B6C4","tomato3", "#FE9929")

mytheme <- theme_classic(base_family = "Helvetica") +
  theme(axis.line.x = element_line(color="black", size = 0.4),
        axis.line.y = element_line(color="black", size = 0.4)) +
  theme(axis.title.x = element_text(size=14),
        axis.text.x  = element_text(size=13),
        axis.title.y = element_text(size=14),
        axis.text.y  = element_text(vjust=0.5, size=13) ) +
  theme(legend.text = element_text(size = 10), legend.title = element_text(size=11), legend.key.size = unit(0.5, "cm")) +
  theme(plot.title = element_text(lineheight=.4, size=13)) +
  theme(plot.margin = unit(c(2.5,1,1,1), "cm")) 

```

## Your submission: `r params$Name` on `r params$Platform` data

```{r, echo=FALSE}
setwd(params$output.directory)

results_file <- paste(params$Name, "_results.RData", sep = '')
load(results_file)
names(all.results) <- c("Transcriptome without reference", "SIRV")

busco_file <- paste(params$Name, "_BUSCO.RData", sep = '')
load(busco_file)

tr_class_file <- paste(params$Name, "_classification.RData", sep = '')
load(tr_class_file)
tr_junc_file <- paste(params$Name, "_junctions.RData", sep = '')
load(tr_junc_file)

SIRV_class_file=paste(params$Name, "_SIRVs_class.RData", sep='')
load(SIRV_class_file)

``` 


### Background

Challenge 3 is evaluated according to a *de novo* long reads-based genome (provided by the LRGASP Consortium) and SIRV Lexogen Set 1 which was included before library preparation.

The LRGASP uses SQANTI3 categories to define evaluating features and metrics for Challenge 3.

### Categories defined by SQANTI3

Due lo the lack of annotation on the *de novo* transcriptome, these categories will only be associated to transcripts that map to SIRV reference sequences. 

- **Full Splice Match (FSM):** Transcripts matching a reference SIRV at all splice junctions.
- **Incomplete Splice Match (ISM):** Transcripts matching consecutive, but not all, splice junctions of the reference SIRV.
- **Novel in Catalog (NIC):**	Transcripts containing new combinations of already annotated splice junctions or novel splice junctions formed from already annotated donors and acceptors. 
- **Novel Not in Catalog (NNC):**	Transcripts using novel donors and/or acceptors.
- **Reference Match (RM):**	FSM transcript with 5´ and 3´ends within 50 nts of the TSS/TTS annotation. This means that a certain SIRV was detected perfectly.

**The rest of the transcripts will be catalogued as Intergenic**

## Evaluation of detected transcripts for Challenge 3  {.tabset .tabset-pills}

These are some definitions used to evaluate the submitted transcriptome:

- **Full Illumina Splice Junction Support:**	Transcripts with all SJ supported by at least one Illumina read.
- **Non-canonical transcripts**	Transcripts with at least one non-canonical junction.
- **Intra-priming:**	It is considered that there is evidence of intra-priming when, in the genomic sequence 20bp downstream the detected TTS, there is at least a 60% of A's.
- **RT-switching:**	Evidence of RT-switching (see SQANTI ref)

### Global metrics

```{r,echo=FALSE}
kable(all.results["Transcriptome without reference"], align = "r")

```

### Length distribution

Length distribution of mapping reads.

```{r, echo=FALSE}
avg_length=as.numeric(all.results$`Transcriptome without reference`["Average length", "Absolute value"])

ggplot(sqanti_data, aes(x=length, fill=structural_category ) )+
  geom_histogram(binwidth=100) +
  mytheme +
  scale_fill_manual(values=myPalette[1])+
  labs(title="Length distribution",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("Length (bp)") +
  geom_vline(xintercept=avg_length)

```

### Minimum SJ coverage

Coverage distribution of the SJ with less coverage in each detected isoform.  

```{r, echo=FALSE}
ggplot(sqanti_data, aes(x=log(min_cov+1), fill=structural_category ) )+
  geom_density(na.rm = T) +
  mytheme +
  scale_fill_manual(values=myPalette[1])+
  labs(title="Minimum SJ coverage",
       subtitle = "Detected isoforms (without SIRVs)") +
  xlab("log(Minimum coverage of SJ +1)")

```

### Coverage comparison: canonical vs non-canonical SJ

Coverage comparison by Illumina reads. Splice Junctions common for several transcripts are counted just once.

```{r, echo=FALSE}
select_cols=c("chrom", "strand", "genomic_start_coord", "genomic_end_coord", "junction_category", "start_site_category", 
                "end_site_category", "splice_site", "canonical", "sample_with_cov", "total_coverage")
unique_SJ=unique(sqanti_data.junc[,select_cols])

ggplot(unique_SJ, aes(x=log(total_coverage + 1), fill=canonical, alpha=0.75))+
  geom_density(na.rm = T) +
  mytheme +
  labs(title="Total coverage of unique SJ",
       subtitle = "Comparison between canonical and non-canonical SJ") +
  xlab("log(Total coverage of SJ +1)")

```

### BUSCO completness results

```{r,echo=FALSE}
kable(busco_results, align = "r")

```

## Evaluation of Spike-Ins (SIRVs) 

The following metrics and definitions apply to SIRV transcripts:

- **SIRV transcript:**	Transcript mapping to a SIRV sequence
- **Reference SIRV (rSIRV):**	Lexogen SIRV model
- **True Positive detections (TP):**	rSIRVs identified as RM 
- **Partial True Positive detections (PTP):**	rSIRVs identified as ISM or FSM_non_RM
- **False Negative (FN):**	rSIRVs without FSM or ISM
- **False Positive (FP):**	NIC + NNC + antisense + fusion SIRV_transcripts
- **Sensitivity:**	TP/rSIRVs
- **Precision:**	RM/ SIRV_transcripts
- **Non_redundant Precision:**	TP/ SIRV_transcripts
- **Positive Detection Rate:**	unique(TP+PTP)/rSIRVs
- **False Discovery Rate:**	(SIRV_transcripts - RM)/SIRV_transcripts
- **Redundancy:**	(FSM + ISM)/unique(TP+PTP)

**ATENTION**
If in this chunk of the evaluation all the results are 0, please, check if the reference genome and/or transcriptome used for building your transcript-model  contain information about spike-ins.


```{r , echo=FALSE}
kable(all.results["SIRV"], align = "r")

```


